### ISSUE TITLE
Auto-cast numbers to strings where needed

### ISSUE THREAD

**Issue by sbliven** [citation-file-format/citation-file-format#91]
pykwalifire currently fails CFF files that use bare numbers for String fields. For instance, `number: 1` fails and ironically must be changed to `number: "1"` (as pointed out by @jspaaks in citation-file-format/schema#9). This is unexpected since most languages would automatically cast the integer to string if needed.

Would it be reasonable to accept `str | int` for string fields (assuming pykwalifire supports type unions somehow)?

**Comment by LourensVeen** [comment]
Yes, YAtiML does unions, but you may still have the issue of the Python code having to check whether it has a number or a string, and to maybe do something different depending on which it is. Better would be to allow strings and integers on the YAML side (so the user wouldn't have to quote a bare number), but convert any numbers to strings while loading, so that the programmer always sees a string. Any strings that are numbers would be converted back to integers on saving, so that the generated YAML doesn't have quotes either and reads naturally.

**Comment by jspaaks** [comment]
Now that we have switched to JSONSchema, we should re-evaluate whether union types for some keys are useful. For example, it is currently (https://github.com/citation-file-format/citation-file-format/blob/38d9b1fb629373f6a3607de64aa3a6e21a76f750/schema.json) invalid to write the following in `CITATION.cff`:

```yaml
version: 1
```
```yaml
version: 1.11
```
```yaml
number: 41
```
```yaml
issue: 11
```

This is a bit unexpected and inconvenient for people particularly those who write CITATION.cff files by hand / unassisted.

I went over the keys and tried to identify a list of keys that could potentially have values that are interpreted as numbers by YAML, which would render them invalid. Here's the list I came up with:

- `#/definitions/post-code`
- `issue`
- `number`
- `scope` (not sure)
- `section`
- `version`

My suggestion would be to have unions using JSONSchema `anyOf`, as follows:

candidates for `str|integer` unions

- `issue`
- `number`

candidates for `str|number` unions

- `#/definitions/post-code`
- `scope`
- `section`
- `version`

**Comment by sdruskat** [comment]
@hainesr This :point_up: discussion / resolution will also be relevant for the currently invalid (see recent discussion) but perhaps valid-again test CFF files.

**Comment by sdruskat** [comment]
Also added this to the 1.2.0 milestone.

**Comment by hainesr** [comment]
If we have `number` as a `string|integer` union, don't we risk having something like this happen:
```yaml
number: '1.0'
```
?

This would be particularly confusing I think.

**Comment by jspaaks** [comment]
https://github.com/citation-file-format/citation-file-format/issues/91#issuecomment-865902434 Yes I agree that is a bit confusing, but not worse than what we have (`number` is currently of type `string`). Also, I don't think many people will be inclined to record their `number` as a `float` with quotes, but if they do, it's not a problem. My guess is that any consumers will cast to `string` anyway. I can't think of a case where tooling would need to do number logic on any of these inputs.

Alternatively, do you think `issue` and `number` should be only `integer`s rather than unions including `integer`?

**Comment by hainesr** [comment]
I suppose my query was really about specifically not allowing `floats` as `floats` for `number` or `issue` versus the others. Sorry I should have been more clear.

If `number` and `issue` can only ever be `integers` then I think limit to that. And if they could be something like `1a` or `Summer` then we do need to allow `strings`. If they could be `1.1` etc then better to allow them to be `number`s than require `floats` to be quoted as strings. Am I making sense?

**Comment by sdruskat** [comment]
My 2¢:

---

**TL;DR:**
- `issue` should be `number|string`
- `number` can be either `string` or `integer|string` (prefer just `string`, which can also be unquoted to be valid YAML)
- `post-code` should be `string`
- `scope` should be `string`
- `section` should be `number|string`
- `version` should be `number|string` (where `1.1.4` (unquoted) will still be a valid string)

---

- Think that `issue` should be `number|string` to cover (stupidly named) issue numbers such as `2.1`. Following Rob here.
- `number` (which is a (library) classification "number") can only be strings AFAIK, by any of the existing classification systems, but also IANAL (I Am Not A Librarian), so I guess we don't lose anything if we allow integers, we just won't be optimally strict.
 
> candidates for `str|number` unions
>     * `#/definitions/post-code`

- I don't see what we gain by allowing `number` for post code. Most codes will simply be (alphanumeric) strings, or perhaps integers (Germany, e.g., 10099 Berlin), but `floats` (analogy for `number` in Ruby and Python)?
 
>     * `scope`

- I don't see a case where this would be a float either, or an integer. If I wanted to supply a section, I'd use `section: 2.1`, see below. I think that scope is more about `scope: "Preface to the 3rd edition"` or `scope: "MyClass.java"` than anything that could be a float. Also as a user even if my scope was a float, I'd go the extra mile and say `scope: "Part 2.1"` than just `scope: 2.1` as I'd be worried about ambiguity.

>     * `section`

- Agree: `str|number` makes sense.

>     * `version`

- Also agree, to make it easier for users to simply say `version: 1.11`.

**Comment by jspaaks** [comment]
I've updated PR #225 to make

1. `issue` of type `number|str`
2. `number` not decided yet, see below
3. `post-code` not decided yet
4. `scope` of type string
5. `section` of type `number|str`
6. `version` of type `number|str`

As for `number`, the README says

- If the work is a conference paper: E.g., the submission number of the paper
- If the work is a grant: The grant number provided by the funding agency.
- If the work is a work of art: E.g., the catalogue number provided by a museum holding the artwork.
- If the work is a report: The report number of a report.
- If the work is a patent: The patent number of the work.
- If the work is a historical work, illuminated manuscript or similar: The codex or folio number of a manuscript, or the library identifier for a manuscript.

Under this definition, I can see `number` being used to record numbers such as `2021.017`. If `number` is of type `string`, writing 
```yaml
number: 2021.017
```
would be valid YAML but would not pass validation because the schema currently says it has to be a string. Here's a table of what styles and number values we need to support, and whether that can be done under 4 scenarios. Based on this, I think `number` should be of type `string|number`.

|   | `integer` | `string` | `string \| integer` | `string \| number` |
| :-- | :-- | :-- | :-- | :-- |
|`number: 2021.017`| :x: | :x: | :x: | :heavy_check_mark: |
|`number: "2021.017"`| :x: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
|`number: ADAS2021.017`| :x: | :heavy_check_mark: | :heavy_check_mark:  | :heavy_check_mark: |
|`number: "ADAS2021.017"`| :x: | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark: |
|`number: 2021017`| :heavy_check_mark:  | :x: | :heavy_check_mark:  | :heavy_check_mark: |
|`number: "2021017"`| :x: | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |

**Comment by jspaaks** [comment]
As for `post-code`, I think we need to support cases where the value consists only of numerals, e.g. `90210`. If the `type` of `post-code` is `string`, the following would be invalid despite being completely unambiguous:

```yaml
post-code: 90210
```

**Comment by hainesr** [comment]
> Based on this, I think `number` should be of type `string|number`

I agree now I've seen the list of things it can encompass. In the UK a grant number can be of the form `EP0000000-X`, for example. And I think the same logic applies to `post-code` given we need to support US, as above, and e.g. UK `M13 9PL`.

**Comment by sdruskat** [comment]
Yup, same here. Should have RTFM!'ed my own docs :blush:.

**Comment by sdruskat** [comment]
From my point-of-view, this can be closed once #225 is merged :rocket:.

**Comment by sdruskat** [comment]
> From my point-of-view, this can be closed once #225 is merged rocket.

Done and done.
